scene/3d/collision_object_3d.cpp

```diff
@@ -505,5 +492,8 @@ CollisionObject3D::CollisionObject3D() {
 }
 
 CollisionObject3D::~CollisionObject3D() {
+	if (_are_collision_shapes_visible()) {
```

Comment:
I'm not sure this condition is reliable, because `is_inside_tree()` might be always false in the destructor and `get_tree()->is_debugging_collisions_hint()` can be changed by script, so this might cause leaks.

Maybe you can use the same logic I did before in the `NOTIFICATION_PREDELETE` case, just counting created debug shapes to know if there's any to destroy.

Reply:
Thinking more about it, I forgot to handle the case when the node is detached from the tree, so this approach is wrong anyway. Now it's handled on enter and exit tree.